synedrion/src/cggmp21/sigma/dec.rs

```diff
@@ -57,101 +95,163 @@ impl<P: SchemeParams> DecProof<P> {
+                };
+                let w = if *e_bit {
+                    ephemeral.beta + secret.y
+                } else {
+                    ephemeral.beta
+                };
+
+                let exponent = if *e_bit {
+                    PublicSigned::one()
+                } else {
+                    PublicSigned::zero()
+                };
+                let nu = secret.rho.to_masked(&ephemeral.r, &exponent);
+
+                DecProofElement {
+                    z: z.to_public(),
+                    w: w.to_public(),
+                    nu,
+                }
+            })
+            .collect::<Vec<_>>();
 
         Self {
             e,
-            cap_s,
-            cap_t,
-            cap_a,
-            gamma,
-            z1,
-            z2,
-            omega,
+            elements: elements.into(),
+            commitments: commitments.into(),
         }
     }
 
     pub fn verify(&self, public: DecPublicInputs<'_, P>, setup: &RPParams<P::Paillier>, aux: &impl Hashable) -> bool {
-        assert_eq!(public.cap_c.public_key(), public.pk0);
-
         let mut reader = XofHasher::new_with_dst(HASH_TAG)
             // commitments
-            .chain(&self.cap_s)
-            .chain(&self.cap_t)
-            .chain(&self.cap_a)
-            .chain(&self.gamma)
+            .chain(&self.commitments)
             // public parameters
             .chain(public.pk0.as_wire())
-            .chain(public.x)
-            .chain(&public.cap_c.to_wire())
+            .chain(&public.cap_k.to_wire())
+            .chain(&public.cap_x)
+            .chain(&public.cap_d.to_wire())
+            .chain(&public.cap_s)
+            .chain(&public.cap_g)
             .chain(&setup.to_wire())
             .chain(aux)
             .finalize_to_reader();
 
         // Non-interactive challenge
-        let e = PublicSigned::from_xof_reader_bounded(&mut reader, &P::CURVE_ORDER);
+        let e = BitVec::from_xof_reader(&mut reader, P::SECURITY_PARAMETER);
 
         if e != self.e {
             return false;
         }
 
-        // enc(z_1, \omega) == A (+) C (*) e
-        if Ciphertext::new_public_with_randomizer_wide(public.pk0, &self.z1, &self.omega)
-            != self.cap_a.to_precomputed(public.pk0) + public.cap_c * &e
-        {
-            return false;
-        }
-
-        // z_1 == \gamma + e x \mod q
-        if scalar_from_wide_signed::<P>(&self.z1) != self.gamma + scalar_from_signed::<P>(&e) * *public.x {
+        if e.bits().len() != self.commitments.len() || e.bits().len() != self.elements.len() {
             return false;
         }
 
-        // s^{z_1} t^{z_2} == T S^e
-        let cap_s = self.cap_s.to_precomputed(setup);
-        let cap_t = self.cap_t.to_precomputed(setup);
-        if setup.commit(&self.z1, &self.z2) != &cap_t * &cap_s.pow(&e) {
-            return false;
+        for ((e_bit, commitment), element) in e
+            .bits()
+            .iter()
+            .cloned()
```

Comment:
Don't think you need this?

Reply:
Can be replaced by `copied()`. If we remove it, it adds a few dereferences.